File: scripts/derive-b52-d12-9-curvature-risk.py
#!/usr/bin/env python3
"""Python producer for the exploratory B52-D12.9-D1 Q30 curvature risk."""
from __future__ import annotations

import argparse
import hashlib
import json
import math
import os
from pathlib import Path

import numpy as np


Q24 = 1 << 24
Q30 = 1 << 30
UINT32_MAX = (1 << 32) - 1


def sha_bytes(value: bytes) -> str:
    return hashlib.sha256(value).hexdigest()


def sha_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def canonical_hash(value: object) -> str:
    encoded = json.dumps(value, sort_keys=True, separators=(",", ":"), allow_nan=False).encode()
    return hashlib.sha256(encoded).hexdigest()


def arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--spec", type=Path, required=True)
    parser.add_argument("--source-root", type=Path, required=True)
    parser.add_argument("--output-root", type=Path, required=True)
    parser.add_argument("--report", type=Path, required=True)
    return parser.parse_args()


def exact_scaled(value: float, scale: int, label: str) -> int:
    scaled = value * scale
    integer = int(scaled)
    if scaled != integer:
        raise RuntimeError(f"non-canonical {label}: {value!r}")
    return integer


def ceil_div(numerator: int, denominator: int) -> int:
    return (numerator + denominator - 1) // denominator


def validate_report(path: Path) -> dict:
    report = json.loads(path.read_text())
    body = {key: value for key, value in report.items() if key != "reportHash"}
    if report.get("reportHash") != canonical_hash(body):
        raise RuntimeError(f"report self-hash mismatch: {path}")
    return report


def load_array(path: Path, expected_sha: str, dtype: str, shape: tuple[int, ...]) -> np.ndarray:
    payload = path.read_bytes()
    if sha_bytes(payload) != expected_sha:
        raise RuntimeError(f"array hash mismatch: {path}")
    return np.frombuffer(payload, dtype=dtype).reshape(shape).copy()


def main() -> None:
    args = arguments()
    if args.output_root.exists() or args.report.exists():
        raise RuntimeError("refusing to overwrite D12.9-D1 Python output")
    spec = json.loads(args.spec.read_text())
    if sha_file(Path(os.path.realpath(os.sys.executable))) != spec["runtime"]["python"]["sha256"] or np.__version__ != spec["runtime"]["python"]["numpy"]:
        raise RuntimeError("Python runtime identity mismatch")
    parent_spec = json.loads((Path.cwd() / spec["sourceEvidence"]["spec"]["uri"]).read_text())
    fixture_by_id = {row["id"]: row for row in parent_spec["fixtures"]}
    output_records: dict[str, object] = {}
    for fixture_id in spec["sourceEvidence"]["fixtures"]:
        fixture = fixture_by_id[fixture_id]
        width, height = fixture["resolution"]
        adapter_dir = args.source_root / "adapters" / fixture_id / "R1"
        consumer_dir = args.source_root / "consumers" / "python" / fixture_id / "R1"
        adapter = validate_report(adapter_dir / "report.json")
        consumer = validate_report(consumer_dir / "report.json")
        previous = load_array(adapter_dir / "arrays" / "previous.rgba32", adapter["arrays"]["previousRgba"]["sha256"], "<f4", (height, width, 4))
        previous_owner = load_array(adapter_dir / "arrays" / "previous-owner.f32", adapter["arrays"]["previousOwner"]["sha256"], "<f4", (height, width))
        vector = load_array(adapter_dir / "arrays" / "vector.xy32", adapter["arrays"]["vector"]["sha256"], "<f4", (height, width, 2))
        radius2 = load_array(consumer_dir / "arrays" / "radius2-interior.u8", consumer["arrays"]["radius2Interior"]["sha256"], "u1", (height, width))
        eligible = np.zeros((height, width), dtype=np.uint8)
        accepted = np.zeros((height, width), dtype=np.uint8)
        risk = np.zeros((height, width, 3), dtype="<u4")
        for y, x in zip(*np.nonzero(radius2)):
            qx = x + float(vector[y, x, 0])
            qy = y - float(vector[y, x, 1])
            x0, y0 = math.floor(qx), math.floor(qy)
            if x0 - 1 < 0 or x0 + 2 >= width or y0 - 1 < 0 or y0 + 2 >= height:
                continue
            owner = previous_owner[y0, x0]
            owner_support = previous_owner[y0 - 1:y0 + 3, x0 - 1:x0 + 3]
            alpha_support = previous[y0 - 1:y0 + 3, x0 - 1:x0 + 3, 3]
            if not np.all(owner_support == owner) or not np.all(alpha_support > np.float32(0.999)):
                continue
            fx = exact_scaled(qx - x0, Q24, "motion fraction x")
            fy = exact_scaled(qy - y0, Q24, "motion fraction y")
            eligible[y, x] = 1
            for channel in range(3):
                color = lambda yy, xx: exact_scaled(float(previous[yy, xx, channel]), Q30, "Q30 RGB")
                mx = max(
                    abs(color(yy, xx - 1) - 2 * color(yy, xx) + color(yy, xx + 1))
                    for yy in (y0, y0 + 1)
                    for xx in (x0, x0 + 1)
                )
                my = max(
                    abs(color(yy - 1, xx) - 2 * color(yy, xx) + color(yy + 1, xx))
                    for xx in (x0, x0 + 1)
                    for yy in (y0, y0 + 1)
                )
                numerator = 2 * (fx * (Q24 - fx) * mx + fy * (Q24 - fy) * my)
                units = ceil_div(numerator, Q24 * Q24) + spec["candidate"]["roundingAllowanceQ30"]
                risk[y, x, channel] = min(units, UINT32_MAX)
            accepted[y, x] = int(int(risk[y, x].max()) <= spec["candidate"]["riskThresholdQ30Inclusive"])
        fixture_dir = args.output_root / fixture_id
        fixture_dir.mkdir(parents=True, exist_ok=False)
        arrays = {}
        for name, array, filename, dtype_name in (
            ("eligible", eligible, "eligible.u8", "uint8"),
            ("accepted", accepted, "accepted.u8", "uint8"),
            ("riskQ30", risk, "risk.q30.u32", "little-endian-uint32"),
        ):
            payload = np.ascontiguousarray(array).tobytes()
            target = fixture_dir / filename
            target.write_bytes(payload)
            arrays[name] = {"uri": str(target), "sha256": sha_bytes(payload), "bytes": len(payload), "shape": list(array.shape), "dtype": dtype_name}
        output_records[fixture_id] = {
            "adapterReport": {"uri": str(adapter_dir / "report.json"), "sha256": sha_file(adapter_dir / "report.json"), "reportHash": adapter["reportHash"]},
            "consumerReport": {"uri": str(consumer_dir / "report.json"), "sha256": sha_file(consumer_dir / "report.json"), "reportHash": consumer["reportHash"]},
            "arrays": arrays,
            "counts": {"radius2": int(radius2.sum()), "eligible": int(eligible.sum()), "accepted": int(accepted.sum())},
        }
    body = {
        "schemaVersion": "bfs.blenderMotionAwareCurvatureRiskProducerReport.v0.1",
        "experimentId": spec["experimentId"],
        "producer": "python",
        "pid": os.getpid(),
        "candidate": spec["candidate"],
        "fixtures": output_records,
        "operationCounts": {"modelCalls": 0, "networkCalls": 0},
    }
    report = {**body, "reportHash": canonical_hash(body)}
    args.report.parent.mkdir(parents=True, exist_ok=True)
    args.report.write_text(json.dumps(report, indent=2, sort_keys=True, allow_nan=False) + "\n")
    print(f"BFS_B52_D129_D1_PYTHON_OK fixtures={len(output_records)}")


if __name__ == "__main__":
    main()

File: scripts/derive-b52-d12-9-curvature-risk.mjs
#!/usr/bin/env node
/** Node producer for the exploratory B52-D12.9-D1 Q30 curvature risk. */
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import process from "node:process";

const Q24 = 1n << 24n;
const Q30 = 1n << 30n;
const UINT32_MAX = (1n << 32n) - 1n;

function arg(name) {
  const index = process.argv.indexOf(name);
  if (index < 0 || index + 1 >= process.argv.length) throw new Error(`missing ${name}`);
  return process.argv[index + 1];
}
function shaBytes(value) { return crypto.createHash("sha256").update(value).digest("hex"); }
function shaFile(filePath) { return shaBytes(fs.readFileSync(filePath)); }
function stable(value) {
  if (Array.isArray(value)) return `[${value.map(stable).join(",")}]`;
  if (value !== null && typeof value === "object") return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${stable(value[key])}`).join(",")}}`;
  if (typeof value === "number" && Number.isFinite(value) && value !== 0 && Math.abs(value) < 1e-4) {
    const [mantissa, exponent] = value.toExponential().split("e");
    const sign = exponent.startsWith("-") ? "-" : "+";
    return `${mantissa}e${sign}${Math.abs(Number(exponent)).toString().padStart(2, "0")}`;
  }
  return JSON.stringify(value);
}
function canonicalHash(value) { return shaBytes(Buffer.from(stable(value))); }
function validateReport(filePath) {
  const report = JSON.parse(fs.readFileSync(filePath, "utf8"));
  const body = Object.fromEntries(Object.entries(report).filter(([key]) => key !== "reportHash"));
  if (report.reportHash !== canonicalHash(body)) throw new Error(`report self-hash mismatch: ${filePath}`);
  return report;
}
function load(filePath, expectedSha) {
  const payload = fs.readFileSync(filePath);
  if (shaBytes(payload) !== expectedSha) throw new Error(`array hash mismatch: ${filePath}`);
  return payload;
}
function exactScaled(value, scale, label) {
  const scaled = value * Number(scale);
  if (!Number.isInteger(scaled)) throw new Error(`non-canonical ${label}: ${value}`);
  return BigInt(scaled);
}
function ceilDiv(numerator, denominator) { return (numerator + denominator - 1n) / denominator; }
function float32(payload) { return new Float32Array(payload.buffer, payload.byteOffset, payload.byteLength / 4); }
function uint8(payload) { return new Uint8Array(payload.buffer, payload.byteOffset, payload.byteLength); }
function f32(array, width, channels, y, x, channel = 0) { return array[(y * width + x) * channels + channel]; }

const specPath = path.resolve(arg("--spec"));
const sourceRoot = path.resolve(arg("--source-root"));
const outputRoot = path.resolve(arg("--output-root"));
const reportPath = path.resolve(arg("--report"));
if (fs.existsSync(outputRoot) || fs.existsSync(reportPath)) throw new Error("refusing to overwrite D12.9-D1 Node output");
const spec = JSON.parse(fs.readFileSync(specPath, "utf8"));
if (shaFile(process.execPath) !== spec.runtime.node.sha256 || process.version !== spec.runtime.node.version) throw new Error("Node runtime identity mismatch");
const parentSpec = JSON.parse(fs.readFileSync(path.resolve(spec.sourceEvidence.spec.uri), "utf8"));
const fixtureById = new Map(parentSpec.fixtures.map((row) => [row.id, row]));
const outputRecords = {};
for (const fixtureId of spec.sourceEvidence.fixtures) {
  const fixture = fixtureById.get(fixtureId);
  const [width, height] = fixture.resolution;
  const adapterDir = path.join(sourceRoot, "adapters", fixtureId, "R1");
  const consumerDir = path.join(sourceRoot, "consumers", "python", fixtureId, "R1");
  const adapterReportPath = path.join(adapterDir, "report.json");
  const consumerReportPath = path.join(consumerDir, "report.json");
  const adapter = validateReport(adapterReportPath);
  const consumer = validateReport(consumerReportPath);
  const previousPayload = load(path.join(adapterDir, "arrays", "previous.rgba32"), adapter.arrays.previousRgba.sha256);
  const ownerPayload = load(path.join(adapterDir, "arrays", "previous-owner.f32"), adapter.arrays.previousOwner.sha256);
  const vectorPayload = load(path.join(adapterDir, "arrays", "vector.xy32"), adapter.arrays.vector.sha256);
  const radius2Payload = load(path.join(consumerDir, "arrays", "radius2-interior.u8"), consumer.arrays.radius2Interior.sha256);
  const previous = float32(previousPayload);
  const previousOwner = float32(ownerPayload);
  const vector = float32(vectorPayload);
  const radius2 = uint8(radius2Payload);
  const eligible = new Uint8Array(width * height);
  const accepted = new Uint8Array(width * height);
  const risk = new Uint32Array(width * height * 3);
  let radius2Count = 0;
  for (let y = 0; y < height; y += 1) for (let x = 0; x < width; x += 1) {
    const pixel = y * width + x;
    if (radius2[pixel] === 0) continue;
    radius2Count += 1;
    const qx = x + f32(vector, width, 2, y, x, 0);
    const qy = y - f32(vector, width, 2, y, x, 1);
    const x0 = Math.floor(qx), y0 = Math.floor(qy);
    if (x0 - 1 < 0 || x0 + 2 >= width || y0 - 1 < 0 || y0 + 2 >= height) continue;
    const owner = f32(previousOwner, width, 1, y0, x0);
    let supportOk = true;
    for (let sy = y0 - 1; sy <= y0 + 2; sy += 1) for (let sx = x0 - 1; sx <= x0 + 2; sx += 1) {
      if (f32(previousOwner, width, 1, sy, sx) !== owner || !(f32(previous, width, 4, sy, sx, 3) > Math.fround(0.999))) supportOk = false;
    }
    if (!supportOk) continue;
    const fx = exactScaled(qx - x0, Q24, "motion fraction x");
    const fy = exactScaled(qy - y0, Q24, "motion fraction y");
    eligible[pixel] = 1;
    for (let channel = 0; channel < 3; channel += 1) {
      const color = (yy, xx) => exactScaled(f32(previous, width, 4, yy, xx, channel), Q30, "Q30 RGB");
      let mx = 0n, my = 0n;
      for (const yy of [y0, y0 + 1]) for (const xx of [x0, x0 + 1]) {
        const value = color(yy, xx - 1) - 2n * color(yy, xx) + color(yy, xx + 1);
        const absolute = value < 0n ? -value : value;
        if (absolute > mx) mx = absolute;
      }
      for (const xx of [x0, x0 + 1]) for (const yy of [y0, y0 + 1]) {
        const value = color(yy - 1, xx) - 2n * color(yy, xx) + color(yy + 1, xx);
        const absolute = value < 0n ? -value : value;
        if (absolute > my) my = absolute;
      }
      const numerator = 2n * (fx * (Q24 - fx) * mx + fy * (Q24 - fy) * my);
      const units = ceilDiv(numerator, Q24 * Q24) + BigInt(spec.candidate.roundingAllowanceQ30);
      risk[pixel * 3 + channel] = Number(units > UINT32_MAX ? UINT32_MAX : units);
    }
    const maximum = Math.max(risk[pixel * 3], risk[pixel * 3 + 1], risk[pixel * 3 + 2]);
    accepted[pixel] = Number(maximum <= spec.candidate.riskThresholdQ30Inclusive);
  }
  const fixtureDir = path.join(outputRoot, fixtureId);
  fs.mkdirSync(fixtureDir, { recursive: true });
  const arrays = {};
  for (const [name, array, filename, dtype] of [
    ["eligible", eligible, "eligible.u8", "uint8"],
    ["accepted", accepted, "accepted.u8", "uint8"],
    ["riskQ30", risk, "risk.q30.u32", "little-endian-uint32"],
  ]) {
    const payload = Buffer.from(array.buffer, array.byteOffset, array.byteLength);
    const target = path.join(fixtureDir, filename);
    fs.writeFileSync(target, payload);
    arrays[name] = { uri: target, sha256: shaBytes(payload), bytes: payload.length, shape: name === "riskQ30" ? [height, width, 3] : [height, width], dtype };
  }
  outputRecords[fixtureId] = {
    adapterReport: { uri: adapterReportPath, sha256: shaFile(adapterReportPath), reportHash: adapter.reportHash },
    consumerReport: { uri: consumerReportPath, sha256: shaFile(consumerReportPath), reportHash: consumer.reportHash },
    arrays,
    counts: { radius2: radius2Count, eligible: eligible.reduce((a, b) => a + b, 0), accepted: accepted.reduce((a, b) => a + b, 0) },
  };
}
const body = {
  schemaVersion: "bfs.blenderMotionAwareCurvatureRiskProducerReport.v0.1",
  experimentId: spec.experimentId,
  producer: "node",
  pid: process.pid,
  candidate: spec.candidate,
  fixtures: outputRecords,
  operationCounts: { modelCalls: 0, networkCalls: 0 },
};
const report = { ...body, reportHash: canonicalHash(body) };
fs.mkdirSync(path.dirname(reportPath), { recursive: true });
fs.writeFileSync(reportPath, `${JSON.stringify(report, null, 2)}\n`);
process.stdout.write(`BFS_B52_D129_D1_NODE_OK fixtures=${Object.keys(outputRecords).length}\n`);
